File: csharp/luhn/Luhn.cs
using System;
using System.Linq;


public static class Luhn
{
    public static bool IsValid(string number)
    {
        if (number.Where(c => !char.IsWhiteSpace(c)).Any(c => !char.IsDigit(c))) return false;

        int DoubleIt(int d)
        {
            if (2 * d > 9) return 2 * d - 9;
            return 2 * d;
        }

        var digits = number
                .Where(char.IsDigit);
        
        if (digits.Count() <= 1) return false;
        
        return digits 
                .Select(c => int.Parse(c.ToString()))
                .Reverse()
                .Select((n, i) => i % 2 == 0 ? n : DoubleIt(n))
                .Sum() % 10 == 0;
     
    }

}

File: fsharp/luhn/Luhn.fs
module Luhn

open System

let valid number = 
    let trimmed = number |> String.filter (Char.IsWhiteSpace >> not)
    if not (trimmed |> String.forall Char.IsDigit) || trimmed.Length < 2 then 
        false
    else 
        trimmed
        |> Seq.map (fun c -> Int32.Parse(c.ToString()))
        |> Seq.rev // reverse it
        |> Seq.mapi (fun i n -> 
            if i % 2 = 1 then  // do the "doubling" only for odd indices
                if n < 5 then 
                    2 * n 
                else 2 * n - 9 
            else n)
        |> Seq.sum 
        |> (%) <| 10  // use "<|" syntax to give 10 as the second argument 
        |> (=) 0